### ISSUE TITLE
allow supplying a custom metadata format parser for compiler

### ISSUE THREAD

**Issue by lennart** [jaspervdj/hakyll#529]
currently hakyll seems to be only able to read metadata as YAML (as far as I understand the code in `Provider/Metadata.hs`)

I want to be able to have hakyll keep source code function documentation in metadata as key value pairs (functionname: functiondocstring) when compiling source code files.

big picture is 
- [x] route to a pattern of haskell source files, compile with a custom `Compiler`
- [ ] then build a map of all functions (docs) found in all files given by a pattern (similar to `buildTags` but not quite)
- [x] then write function references within markdown pages as standard yaml metadata list field
- [x] build a map of those function references
- [ ] match them against the function docs by their identifiers (function names)
- [ ] render the markdown page with content & a list of rendered functionname/functiondocstrings

So in order to do that (as the docstrings do not appear in single files and I cannot find another way to do the above) I was thinking of a custom Metadata parser (I can provide) that essentially parses haskell source code into an AST and extracts the function names and docstrings into a `Metadata` object.

How would that be possible? I am not hesitant to put this into a contrib module, as I fear this is a rather non-standard feature request.

However right now, I think I would need to fork hakyll in order to do that, and I would rather not.

suggestions on how to proceed?

kindly,

lennart

**Comment by jaspervdj** [comment]
I don't think you necessarily need to do this through metadata, I think you can approach it from a different angle as well.

What I suggest is the following flow:

1. Compile all source files to a custom datatype (e.g. `DocMap`).  Use a `version "docmap"` here to distinguish it.
2. Add a compiler which loads all `DocMap`s produced in (1) and `mconcat`s them together to get a "global" `DocMap`.
3. Compile all source files "normally" but load the global `DocMap` so you can use that to render the table of context and so on.
